File: tricks/recitation.ts
/**
 * Trick Q: Recitation / Goal Reinforcement
 *
 * Periodically injects task summaries (plans, todos, goals) into the END
 * of the context to combat "lost in the middle" attention issues.
 *
 * Problem: In long conversations (50+ tool calls), the model "forgets"
 * the original goal because it's buried in the middle of the context.
 *
 * Solution: Periodically "recite" the current goal/plan at the context end,
 * pushing it into the model's recent attention span.
 *
 * @example
 * ```typescript
 * import { createRecitationManager } from './recitation';
 *
 * const recitation = createRecitationManager({
 *   frequency: 5,  // Every 5 iterations
 *   sources: ['plan', 'todo', 'goal'],
 * });
 *
 * // In agent loop
 * const messages = recitation.injectIfNeeded(
 *   currentMessages,
 *   { plan, todos, goal, iteration: 15 }
 * );
 * ```
 */

// =============================================================================
// TYPES
// =============================================================================

/**
 * Sources of content to recite.
 */
export type RecitationSource = 'plan' | 'todo' | 'goal' | 'memory' | 'custom';

/**
 * Configuration for recitation.
 */
export interface RecitationConfig {
  /** Inject recitation every N iterations (default: 5) */
  frequency: number;

  /** Sources to include in recitation */
  sources: RecitationSource[];

  /** Maximum tokens for recitation block (default: 500) */
  maxTokens?: number;

  /** Custom recitation builder */
  customBuilder?: (state: RecitationState) => string | null;

  /** Whether to track injection history */
  trackHistory?: boolean;
}

/**
 * Current state for building recitation.
 */
export interface RecitationState {
  /** Current iteration number */
  iteration: number;

  /** Original user goal/task */
  goal?: string;

  /** Current plan with tasks */
  plan?: PlanState;

  /** Current todo list */
  todos?: TodoItem[];

  /** Relevant memory items */
  memories?: string[];

  /** Files currently being worked on */
  activeFiles?: string[];

  /** Recent errors encountered */
  recentErrors?: string[];

  /** Custom state values */
  custom?: Record<string, unknown>;
}

/**
 * Plan state.
 */
export interface PlanState {
  description: string;
  tasks: PlanTask[];
  currentTaskIndex: number;
}

/**
 * Task in a plan.
 */
export interface PlanTask {
  id: string;
  description: string;
  status: 'pending' | 'in_progress' | 'completed' | 'failed';
}

/**
 * Todo item.
 */
export interface TodoItem {
  content: string;
  status: 'pending' | 'in_progress' | 'completed';
}

/**
 * Message for injection.
 */
export interface RecitationMessage {
  role: 'system' | 'user' | 'assistant';
  content: string;
}

/**
 * Recitation history entry.
 */
export interface RecitationEntry {
  iteration: number;
  timestamp: string;
  content: string;
  sources: RecitationSource[];
}

/**
 * Events emitted by recitation manager.
 */
export type RecitationEvent =
  | { type: 'recitation.injected'; iteration: number; sources: RecitationSource[] }
  | { type: 'recitation.skipped'; iteration: number; reason: string }
  | { type: 'recitation.built'; content: string; tokenEstimate: number };

export type RecitationEventListener = (event: RecitationEvent) => void;

// =============================================================================
// RECITATION MANAGER
// =============================================================================

/**
 * Manages periodic goal/plan recitation.
 */
export class RecitationManager {
  private config: Required<RecitationConfig>;
  private history: RecitationEntry[] = [];
  private lastInjectionIteration = 0;
  private listeners: RecitationEventListener[] = [];

  constructor(config: RecitationConfig) {
    this.config = {
      frequency: config.frequency,
      sources: config.sources,
      maxTokens: config.maxTokens ?? 500,
      customBuilder: config.customBuilder ?? (() => null),
      trackHistory: config.trackHistory ?? true,
    };
  }

  /**
   * Check if recitation should be injected at this iteration.
   */
  shouldInject(iteration: number): boolean {
    // Always inject on first iteration
    if (iteration === 1) return true;

    // Inject every N iterations
    const iterationsSinceLastInjection = iteration - this.lastInjectionIteration;
    return iterationsSinceLastInjection >= this.config.frequency;
  }

  /**
   * Build recitation content from state.
   */
  buildRecitation(state: RecitationState): string | null {
    const parts: string[] = [];
    const usedSources: RecitationSource[] = [];

    // Goal
    if (this.config.sources.includes('goal') && state.goal) {
      parts.push(`**Goal**: ${state.goal}`);
      usedSources.push('goal');
    }

    // Plan progress
    if (this.config.sources.includes('plan') && state.plan) {
      const planSummary = this.buildPlanSummary(state.plan);
      if (planSummary) {
        parts.push(planSummary);
        usedSources.push('plan');
      }
    }

    // Todo list
    if (this.config.sources.includes('todo') && state.todos?.length) {
      const todoSummary = this.buildTodoSummary(state.todos);
      if (todoSummary) {
        parts.push(todoSummary);
        usedSources.push('todo');
      }
    }

    // Memory context
    if (this.config.sources.includes('memory') && state.memories?.length) {
      const memorySummary = this.buildMemorySummary(state.memories);
      if (memorySummary) {
        parts.push(memorySummary);
        usedSources.push('memory');
      }
    }

    // Custom content
    if (this.config.sources.includes('custom')) {
      const customContent = this.config.customBuilder(state);
      if (customContent) {
        parts.push(customContent);
        usedSources.push('custom');
      }
    }

    // Active files (always include if present)
    if (state.activeFiles?.length) {
      parts.push(`**Active files**: ${state.activeFiles.join(', ')}`);
    }

    // Recent errors (warn about them)
    if (state.recentErrors?.length) {
      parts.push(`**Recent errors**: ${state.recentErrors.slice(-2).join('; ')}`);
    }

    if (parts.length === 0) {
      return null;
    }

    const content = parts.join('\n\n');

    // Truncate if too long
    const tokenEstimate = Math.ceil(content.length / 4);
    if (tokenEstimate > this.config.maxTokens) {
      const truncated = this.truncateContent(content, this.config.maxTokens);
      this.emit({ type: 'recitation.built', content: truncated, tokenEstimate: this.config.maxTokens });
      return truncated;
    }

    this.emit({ type: 'recitation.built', content, tokenEstimate });
    return content;
  }

  /**
   * Inject recitation into messages if needed.
   * Returns the modified message array.
   */
  injectIfNeeded<T extends RecitationMessage>(
    messages: T[],
    state: RecitationState
  ): T[] {
    if (!this.shouldInject(state.iteration)) {
      this.emit({
        type: 'recitation.skipped',
        iteration: state.iteration,
        reason: 'not_due',
      });
      return messages;
    }

    const recitationContent = this.buildRecitation(state);
    if (!recitationContent) {
      this.emit({
        type: 'recitation.skipped',
        iteration: state.iteration,
        reason: 'no_content',
      });
      return messages;
    }

    // Create recitation message
    const recitationMessage = {
      role: 'system' as const,
      content: `[Current Status - Iteration ${state.iteration}]\n${recitationContent}`,
    } as T;

    // Track injection
    this.lastInjectionIteration = state.iteration;

    if (this.config.trackHistory) {
      this.history.push({
        iteration: state.iteration,
        timestamp: new Date().toISOString(),
        content: recitationContent,
        sources: this.config.sources.filter(s =>
          recitationContent.toLowerCase().includes(s.toLowerCase())
        ),
      });
    }

    this.emit({
      type: 'recitation.injected',
      iteration: state.iteration,
      sources: this.config.sources,
    });

    // Inject at the END of messages (before the final user message if present)
    const lastUserIndex = this.findLastUserMessageIndex(messages);
    if (lastUserIndex >= 0) {
      // Insert before the last user message
      return [
        ...messages.slice(0, lastUserIndex),
        recitationMessage,
        ...messages.slice(lastUserIndex),
      ];
    }

    // No user message at end, just append
    return [...messages, recitationMessage];
  }

  /**
   * Force recitation injection regardless of frequency.
   */
  forceInject<T extends RecitationMessage>(
    messages: T[],
    state: RecitationState
  ): T[] {
    const originalFrequency = this.config.frequency;
    this.config.frequency = 0; // Force injection
    const result = this.injectIfNeeded(messages, state);
    this.config.frequency = originalFrequency;
    return result;
  }

  /**
   * Get injection history.
   */
  getHistory(): RecitationEntry[] {
    return [...this.history];
  }

  /**
   * Clear history.
   */
  clearHistory(): void {
    this.history = [];
    this.lastInjectionIteration = 0;
  }

  /**
   * Update configuration.
   */
  updateConfig(updates: Partial<RecitationConfig>): void {
    Object.assign(this.config, updates);
  }

  /**
   * Subscribe to events.
   */
  on(listener: RecitationEventListener): () => void {
    this.listeners.push(listener);
    return () => {
      const idx = this.listeners.indexOf(listener);
      if (idx >= 0) this.listeners.splice(idx, 1);
    };
  }

  // Internal methods

  private buildPlanSummary(plan: PlanState): string {
    const completedCount = plan.tasks.filter(t => t.status === 'completed').length;
    const totalCount = plan.tasks.length;
    const currentTask = plan.tasks[plan.currentTaskIndex];

    const lines = [
      `**Plan Progress**: ${completedCount}/${totalCount} tasks completed`,
    ];

    if (currentTask) {
      lines.push(`**Current Task**: ${currentTask.description}`);
    }

    // Show next pending tasks (max 2)
    const pendingTasks = plan.tasks
      .filter(t => t.status === 'pending')
      .slice(0, 2);

    if (pendingTasks.length > 0) {
      lines.push('**Next**: ' + pendingTasks.map(t => t.description).join('; '));
    }

    return lines.join('\n');
  }

  private buildTodoSummary(todos: TodoItem[]): string {
    const pending = todos.filter(t => t.status === 'pending');
    const inProgress = todos.filter(t => t.status === 'in_progress');
    const completed = todos.filter(t => t.status === 'completed');

    const lines = [
      `**Todo Status**: ${completed.length} done, ${inProgress.length} active, ${pending.length} pending`,
    ];

    if (inProgress.length > 0) {
      lines.push('**In Progress**: ' + inProgress.map(t => t.content).join('; '));
    }

    if (pending.length > 0 && pending.length <= 3) {
      lines.push('**Remaining**: ' + pending.map(t => t.content).join('; '));
    } else if (pending.length > 3) {
      lines.push(
        '**Next**: ' + pending.slice(0, 3).map(t => t.content).join('; ') +
        ` (+${pending.length - 3} more)`
      );
    }

    return lines.join('\n');
  }

  private buildMemorySummary(memories: string[]): string {
    if (memories.length === 0) return '';

    const maxMemories = 3;
    const shown = memories.slice(0, maxMemories);

    let summary = '**Relevant Context**: ' + shown.join('; ');
    if (memories.length > maxMemories) {
      summary += ` (+${memories.length - maxMemories} more)`;
    }

    return summary;
  }

  private truncateContent(content: string, maxTokens: number): string {
    const maxChars = maxTokens * 4;
    if (content.length <= maxChars) return content;

    return content.slice(0, maxChars - 20) + '\n...[truncated]';
  }

  private findLastUserMessageIndex<T extends RecitationMessage>(messages: T[]): number {
    for (let i = messages.length - 1; i >= 0; i--) {
      if (messages[i].role === 'user') {
        return i;
      }
    }
    return -1;
  }

  private emit(event: RecitationEvent): void {
    for (const listener of this.listeners) {
      try {
        listener(event);
      } catch {
        // Ignore listener errors
      }
    }
  }
}

// =============================================================================
// FACTORY FUNCTIONS
// =============================================================================

/**
 * Create a recitation manager.
 *
 * @example
 * ```typescript
 * const recitation = createRecitationManager({
 *   frequency: 5,  // Every 5 iterations
 *   sources: ['goal', 'plan', 'todo'],
 *   maxTokens: 500,
 * });
 *
 * // In agent loop
 * const enrichedMessages = recitation.injectIfNeeded(messages, {
 *   iteration: currentIteration,
 *   goal: 'Implement user authentication',
 *   plan: currentPlan,
 *   todos: currentTodos,
 * });
 * ```
 */
export function createRecitationManager(
  config: RecitationConfig
): RecitationManager {
  return new RecitationManager(config);
}

// =============================================================================
// UTILITIES
// =============================================================================

/**
 * Quick helper to build a recitation block without full manager.
 */
export function buildQuickRecitation(state: RecitationState): string {
  const parts: string[] = [];

  if (state.goal) {
    parts.push(`Goal: ${state.goal}`);
  }

  if (state.plan) {
    const completed = state.plan.tasks.filter(t => t.status === 'completed').length;
    parts.push(`Progress: ${completed}/${state.plan.tasks.length} tasks`);

    const current = state.plan.tasks[state.plan.currentTaskIndex];
    if (current) {
      parts.push(`Current: ${current.description}`);
    }
  }

  if (state.todos?.length) {
    const pending = state.todos.filter(t => t.status !== 'completed').length;
    parts.push(`Todos: ${pending} remaining`);
  }

  return parts.join(' | ');
}

/**
 * Calculate optimal recitation frequency based on context size.
 * Longer contexts need more frequent recitation.
 */
export function calculateOptimalFrequency(contextTokens: number): number {
  if (contextTokens < 10000) return 10;  // Light context
  if (contextTokens < 30000) return 7;   // Medium context
  if (contextTokens < 60000) return 5;   // Heavy context
  return 3;  // Very heavy context - recite often
}

/**
 * Format recitation history for debugging.
 */
export function formatRecitationHistory(history: RecitationEntry[]): string {
  if (history.length === 0) {
    return 'No recitation history.';
  }

  const lines = ['Recitation History:', ''];

  for (const entry of history.slice(-5)) {
    lines.push(`[Iteration ${entry.iteration}] (${entry.sources.join(', ')})`);
    lines.push(`  ${entry.content.slice(0, 100)}${entry.content.length > 100 ? '...' : ''}`);
    lines.push('');
  }

  return lines.join('\n');
}

File: attocode/attocode_py/src/attocode/tricks/recitation.py
"""Goal recitation for maintaining focus (Trick Q).

Periodically injects goal/progress summaries into the message
stream to prevent the agent from losing track of objectives.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Callable

from attocode.integrations.utilities.token_estimate import estimate_tokens


@dataclass
class PlanTask:
    """A task in a plan."""

    id: str = ""
    description: str = ""
    status: str = "pending"  # pending, in_progress, completed, failed


@dataclass
class PlanState:
    """State of a plan."""

    description: str = ""
    tasks: list[PlanTask] = field(default_factory=list)
    current_task_index: int = 0


@dataclass
class TodoItem:
    """A todo item."""

    content: str = ""
    status: str = "pending"


@dataclass
class RecitationState:
    """Current state for building recitations."""

    iteration: int = 0
    goal: str | None = None
    plan: PlanState | None = None
    todos: list[TodoItem] | None = None
    memories: list[str] | None = None
    active_files: list[str] | None = None
    recent_errors: list[str] | None = None
    custom: dict[str, str] | None = None


@dataclass
class RecitationConfig:
    """Configuration for recitation injection."""

    frequency: int = 5  # Inject every N iterations
    sources: list[str] = field(default_factory=lambda: ["plan", "todo", "goal"])
    max_tokens: int = 500
    track_history: bool = True


@dataclass
class RecitationEntry:
    """A recorded recitation injection."""

    iteration: int
    content: str
    sources: list[str] = field(default_factory=list)


RecitationEventListener = Callable[[str, dict[str, Any]], None]


class RecitationManager:
    """Manages periodic goal/progress recitation.

    Injects system messages at configurable intervals to
    keep the agent focused on its objectives.
    """

    def __init__(self, config: RecitationConfig | None = None) -> None:
        self._config = config or RecitationConfig()
        self._last_injection_iteration = 0
        self._history: list[RecitationEntry] = []
        self._listeners: list[RecitationEventListener] = []

    def should_inject(self, iteration: int) -> bool:
        """Check if recitation should be injected at this iteration."""
        if iteration <= 1:
            return True
        return (iteration - self._last_injection_iteration) >= self._config.frequency

    def build_recitation(self, state: RecitationState) -> str | None:
        """Build recitation content from current state.

        Returns None if there's nothing meaningful to recite.
        """
        sections: list[str] = []

        # Goal
        if "goal" in self._config.sources and state.goal:
            sections.append(f"Goal: {state.goal}")

        # Plan progress
        if "plan" in self._config.sources and state.plan:
            plan = state.plan
            total = len(plan.tasks)
            completed = sum(1 for t in plan.tasks if t.status == "completed")
            sections.append(f"Plan: {completed}/{total} tasks complete")

            # Show next pending tasks (up to 2)
            pending = [t for t in plan.tasks if t.status == "pending"]
            for task in pending[:2]:
                sections.append(f"  Next: {task.description}")

        # Todos
        if "todo" in self._config.sources and state.todos:
            pending_todos = [t for t in state.todos if t.status == "pending"]
            if pending_todos:
                sections.append(f"Todos: {len(pending_todos)} remaining")
                for todo in pending_todos[:3]:
                    sections.append(f"  - {todo.content}")

        # Memories
        if "memory" in self._config.sources and state.memories:
            for mem in state.memories[:3]:
                sections.append(f"Remember: {mem}")

        # Recent errors
        if state.recent_errors:
            for err in state.recent_errors[-2:]:
                sections.append(f"Recent error: {err}")

        # Custom
        if state.custom:
            for k, v in state.custom.items():
                sections.append(f"{k}: {v}")

        if not sections:
            return None

        content = "\n".join(sections)

        # Truncate to max tokens
        max_chars = int(self._config.max_tokens * 3.5)
        if len(content) > max_chars:
            content = content[:max_chars] + "..."

        return content

    def inject_if_needed(
        self,
        messages: list[dict[str, Any]],
        state: RecitationState,
    ) -> list[dict[str, Any]]:
        """Inject recitation into messages if needed.

        Returns a new list with the recitation message inserted
        before the last user message.
        """
        if not self.should_inject(state.iteration):
            self._emit("recitation.skipped", {"iteration": state.iteration})
            return messages

        content = self.build_recitation(state)
        if content is None:
            return messages

        # Wrap in status header
        full_content = f"[Current Status - Iteration {state.iteration}]\n{content}"

        recitation_msg: dict[str, Any] = {
            "role": "system",
            "content": full_content,
        }

        # Insert before last user message
        result = list(messages)
        insert_idx = len(result)
        for i in range(len(result) - 1, -1, -1):
            if result[i].get("role") == "user":
                insert_idx = i
                break
        result.insert(insert_idx, recitation_msg)

        self._last_injection_iteration = state.iteration
        self._emit("recitation.injected", {"iteration": state.iteration})

        # Track history
        if self._config.track_history:
            self._history.append(RecitationEntry(
                iteration=state.iteration,
                content=content,
                sources=list(self._config.sources),
            ))

        return result

    def force_inject(
        self,
        messages: list[dict[str, Any]],
        state: RecitationState,
    ) -> list[dict[str, Any]]:
        """Force recitation injection regardless of frequency."""
        saved = self._config.frequency
        self._config.frequency = 0
        result = self.inject_if_needed(messages, state)
        self._config.frequency = saved
        return result

    def get_history(self) -> list[RecitationEntry]:
        """Get recitation history."""
        return list(self._history)

    def clear_history(self) -> None:
        """Clear recitation history."""
        self._history.clear()

    def update_config(self, **kwargs: Any) -> None:
        """Update configuration."""
        for key, value in kwargs.items():
            if hasattr(self._config, key):
                setattr(self._config, key, value)

    def on(self, listener: RecitationEventListener) -> Callable[[], None]:
        """Subscribe to recitation events."""
        self._listeners.append(listener)

        def unsubscribe() -> None:
            try:
                self._listeners.remove(listener)
            except ValueError:
                pass

        return unsubscribe

    def _emit(self, event: str, data: dict[str, Any]) -> None:
        for listener in self._listeners:
            try:
                listener(event, data)
            except Exception:
                pass


def build_quick_recitation(state: RecitationState) -> str:
    """Build a compact one-line recitation."""
    parts: list[str] = []

    if state.goal:
        parts.append(f"Goal: {state.goal}")

    if state.plan:
        completed = sum(1 for t in state.plan.tasks if t.status == "completed")
        total = len(state.plan.tasks)
        parts.append(f"Progress: {completed}/{total}")

    if state.todos:
        pending = sum(1 for t in state.todos if t.status == "pending")
        parts.append(f"Todos: {pending} remaining")

    return " | ".join(parts)


def calculate_optimal_frequency(context_tokens: int) -> int:
    """Calculate optimal recitation frequency based on context size."""
    if context_tokens < 10_000:
        return 10
    elif context_tokens < 30_000:
        return 7
    elif context_tokens < 60_000:
        return 5
    else:
        return 3
